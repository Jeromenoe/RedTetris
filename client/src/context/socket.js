import React from 'react';
import socketio from "socket.io-client";


const localHost = "http://localhost:3001"; 
const herokuHost = "https://redtetris42-heroku.herokuapp.com"
const socketURL = process.env.NODE_ENV === "production" ? herokuHost : localHost

export const socket = socketio.connect(socketURL);
export const SocketContext = React.createContext();