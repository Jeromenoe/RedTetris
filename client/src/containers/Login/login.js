import React, { useContext, useState } from 'react';
import { UserLoginContainer, UserInputContainer} from './styles'
import { updateData } from '../../actions/game'
import { SocketContext } from '../../context/socket';
import store from '../../store/index';

export const Login = () => {
    let [errorLog, setErrorLog] = useState(null); 
    const socket = useContext(SocketContext);

    const loginServer = (e) => {
        let playerName = document.getElementById("playerNameId").value;
        let gameName = document.getElementById("gameNameId").value;
        socket.emit('joinGame', {roomName: gameName, playerName: playerName}, (response) => {
            if (response === 'ok') {
                const me = store.getState().me;
                let clonedMe = {
                    ...me,
                    name: playerName,
                };
                store.dispatch(updateData({me: clonedMe}));
            } else {
                setErrorLog(response);
            }
        });
    }
    getUrlInfos(socket);
    return (
      <UserLoginContainer>
        <h1 className="redTetrisTitle">Red Tetris</h1>
        <UserInputContainer>
            <input type='text' id="playerNameId" className="userInfosTextField" placeholder="User Name"/>
            {errorLog && errorLog.playerName && <p className="error-login">{errorLog.playerName}</p> }
        </UserInputContainer>
        <UserInputContainer>
            <input type='text' id="gameNameId" className="userInfosTextField" placeholder="Game Name"/>
            {errorLog && errorLog.gameName && <p className="error-login">{errorLog.gameName}</p>}
        </UserInputContainer>
        <button className="PlayButton" onClick={loginServer}>PLAY</button>
      </UserLoginContainer>
      
    )
  }

export const getUrlInfos = (socket) => {
    const gameName = retrieveGameNameFromUrl(window.location.href);
    const playerName = retrievePlayerNameFromUrl(window.location.href);

    if (!gameName || !playerName)
        return;
    socket.emit('joinGame', {roomName: gameName, playerName: playerName}, (response) => {
        if (response === 'ok') {
            const me = store.getState().me;
            const game = store.getState().game;
            let clonedMe = {
                ...me,
                name: playerName,
            };
            let clonedGame = {
                ...game,
                name: gameName,
            };
            store.dispatch(updateData({ me: clonedMe, game: clonedGame }))
        }
    });
}

function retrieveGameNameFromUrl(url) {
    const indexOfHashUrl = url.indexOf('#');
    const indexOfOpeningBracket = url.indexOf('[');

    if (indexOfHashUrl === -1 || indexOfHashUrl > indexOfOpeningBracket)
        return null;
    const roomName = url.slice(indexOfHashUrl + 1, indexOfOpeningBracket);
    if (!roomName)
        return null;
    return roomName;
}

function retrievePlayerNameFromUrl(url) {
    const indexOfOpeningBracket = url.indexOf('[');
    const indexOfClosingBracket = url.indexOf(']');

    if (indexOfOpeningBracket === -1 || indexOfClosingBracket === -1
        || indexOfOpeningBracket > indexOfClosingBracket) {
        return null;
    }
    const playerName = url.slice(indexOfOpeningBracket + 1, indexOfClosingBracket);
    if (!playerName)
        return null;
    return playerName;
}