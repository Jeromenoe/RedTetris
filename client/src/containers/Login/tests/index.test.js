import React from 'react';
import {Login} from '../login';

import configureStore from "redux-mock-store";
import {Provider} from 'react-redux';
import toJson from "enzyme-to-json";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';

import { SocketContext, socket } from "../../../context/socket"


configure({adapter: new Adapter()});


beforeEach(() => {
	// Avoid `attachTo: document.body` Warning
	const div = document.createElement('div');
	div.setAttribute('id', 'container');
	document.body.appendChild(div);
  });
  
  afterEach(() => {
	const div = document.getElementById('container');
	if (div) {
	  document.body.removeChild(div);
	}
  });

describe('Testing Login Container', () => {
	test('Check render', async () => {
		const mockStore = configureStore();
		const initialState = {
			game: {
				name: null,
				isStarted: false,
				isOver: false,
				opponents: [],
				players: [],
				removeOwnIndestructibleLine: false,
				difficulty: 2,
				nbOfNextTetromino: 3,
			},
			me: {
				name: null,
				hasLost: false,
				stage: [[]],
				nextPieces: [],
				rows: 0,
				level: 0,
				score: 0
			}
		};
		const store = mockStore(initialState);

		const wrapper = mount(
		<Provider store={store}><SocketContext.Provider value={socket}><Login /></SocketContext.Provider></Provider>, 
		{ attachTo: document.getElementById('container') });
		expect(toJson(wrapper)).toMatchSnapshot();

		wrapper.find('.PlayButton').simulate('click');

	  });

	  test('Check good url', async () => {
		const mockStore = configureStore();
		const initialState = {
			game: {
				name: null,
				isStarted: false,
				isOver: false,
				opponents: [],
				players: [],
				removeOwnIndestructibleLine: false,
				difficulty: 2,
				nbOfNextTetromino: 3,
			},
			me: {
				name: null,
				hasLost: false,
				stage: [[]],
				nextPieces: [],
				rows: 0,
				level: 0,
				score: 0
			}
		};
		const store = mockStore(initialState);

		delete window.location;
		window.location = {}
		window.location.href = 'http://localhost:3000#game[toto]';

		const wrapper = mount(
		<Provider store={store}><SocketContext.Provider value={socket}><Login /></SocketContext.Provider></Provider>, 
		{ attachTo: document.getElementById('container') });
		expect(toJson(wrapper)).toMatchSnapshot();

	  });
});