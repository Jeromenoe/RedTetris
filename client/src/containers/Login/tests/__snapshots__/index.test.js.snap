// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login Container Check good url 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Login>
    <styled.div>
      <div
        className="sc-gtsrHT bpUZGv"
      >
        <h1
          className="redTetrisTitle"
        >
          Red Tetris
        </h1>
        <styled.div>
          <div
            className="sc-bdnxRM gcfnXe"
          >
            <input
              className="userInfosTextField"
              id="playerNameId"
              placeholder="User Name"
              type="text"
            />
          </div>
        </styled.div>
        <styled.div>
          <div
            className="sc-bdnxRM gcfnXe"
          >
            <input
              className="userInfosTextField"
              id="gameNameId"
              placeholder="Game Name"
              type="text"
            />
          </div>
        </styled.div>
        <button
          className="PlayButton"
          onClick={[Function]}
        >
          PLAY
        </button>
      </div>
    </styled.div>
  </Login>
</Provider>
`;

exports[`Testing Login Container Check render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Login>
    <styled.div>
      <div
        className="sc-gtsrHT bpUZGv"
      >
        <h1
          className="redTetrisTitle"
        >
          Red Tetris
        </h1>
        <styled.div>
          <div
            className="sc-bdnxRM gcfnXe"
          >
            <input
              className="userInfosTextField"
              id="playerNameId"
              placeholder="User Name"
              type="text"
            />
          </div>
        </styled.div>
        <styled.div>
          <div
            className="sc-bdnxRM gcfnXe"
          >
            <input
              className="userInfosTextField"
              id="gameNameId"
              placeholder="Game Name"
              type="text"
            />
          </div>
        </styled.div>
        <button
          className="PlayButton"
          onClick={[Function]}
        >
          PLAY
        </button>
      </div>
    </styled.div>
  </Login>
</Provider>
`;
